### DISCUSSION TITLE
Serialization/Deserialization of an instance of a class to a file

### DISCUSSION THREAD

**Discussion by istinnstudio** [activej/activej#202] [Q&A] [answered]
Is it possible for activej serializer to serialize an instance of a class to a file and then de-serialize this file? Is there an example that can do similar functions like the native (de)serialization?
If we mark the serializable fields with the corresponding annotation in the class, what is the next step to save/load the byte array?
What we can do to replicate/emulate the native code?
Thanks

EXAMPLE: file i/o with native serialize/deserialize

   ```
public class ClassX implements Serializable{

//several fields and methods

private static final void saveInstanceX(ClassX instance){

      String pathtoconfig=ClassX.getLocalDir().getPath()+"/"+CLASSXINSTANCEFILENAME;
    System.out.println("Saving an instance of native serialization file : "+pathtoconfig);
    FileOutputStream fos;
    ObjectOutputStream oos;
    File file=new File(pathtoconfig);
    try{
      fos=new FileOutputStream(file);
      oos=new ObjectOutputStream(fos);
      oos.writeObject(instance);
      oos.close();
    }catch(IOException x){
      System.out.println("Saving of native serialization instance failed.");
      x.printStackTrace();}
   }
   
     private static final ClassX loadInstanceX(){
     
    String pathtoconfig=ClassX.getLocalDir().getPath()+"/"+CLASSXINSTANCEFILENAME;
    System.out.println("Loading instance : "+pathtoconfig);
    ClassX instance=null;
    try{
      FileInputStream a=new FileInputStream(pathtoconfig);
      ObjectInputStream b=new ObjectInputStream(a);
      instance=(ClassX)b.readObject();
      b.close();
    }catch(Exception e){
      System.out.println("Loading of native serialization instance failed.");}
    return instance;
     }
}
```

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @0248991
In order to serialize/deserialize an instance of a class you need to create a corresponding serializer for that class. In order to do that you need to mark serialized fields with[ `@Serialize`](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/annotations/Serialize.java) annotations and then you would have to construct a serializer for that class using a [`SerializerBuilder`](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/SerializerBuilder.java). You would get an instance of a [`BinarySerializer`](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/BinarySerializer.java) for that class that is capable of serializing class instances into a byte array. You could then save that byte array to a file if you wish to do that. 

In order to deserialize a class instance from a file, you can use the same `BinarySerializer` that you used for the serialization. You would need to read file content to a byte array and then use the `BinarySerializer#decode` method on that byte array.

Please, take a look at [this example](https://activej.io/serializer/examples#simple-object-serialization) that demonstrates how to serialize a class to a byte array.
